master/buildbot/process/buildrequestdistributor.py

```diff
@@ -0,0 +1,1004 @@
+# coding=utf-8
+# This file is part of Buildbot.  Buildbot is free software: you can
+# redistribute it and/or modify it under the terms of the GNU General Public
+# License as published by the Free Software Foundation, version 2.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright Buildbot Team Members
+
+
```

Comment:
There should be a section in the developer documentation which describes the high level algorithm used for scheduling builds. How the slaves are picked, how the builds are prioritized, etc.

And we should add a reference to that section in this file.